File: go/Vector/Vector.go
package Vector

import (
	"errors"
	"fmt"
	"math"
)

type Vector struct {
	Components []float32
	Size       int
}

func (v Vector) String() string {
	return fmt.Sprintf("%v", v.Components)
}

func NewVector(components []float32) Vector {
	return Vector{
		Size:       len(components),
		Components: components,
	}
}

func (v *Vector) Scale(scalar float32) {
	for i := 0; i < len(v.Components); i++ {
		v.Components[i] *= scalar
	}
}

func (v *Vector) AtRef(index int) (*float32, error) {
	if index < 0 || index >= len(v.Components) {
		var zero float32 = 0
		return &zero, errors.New("OutOfBounds: index is out of bounds from Vector")
	}
	return &v.Components[index], nil
}

func (v *Vector) At(index int) (float32, error) {
	if index < 0 || index >= len(v.Components) {
		return 0.0, errors.New("OutOfBounds: index is out of bounds from Vector")
	}
	return v.Components[index], nil
}

func (v *Vector) Magnitude() float32 {
	dot_product, _ := Dot(v, v)
	return float32(math.Sqrt(float64(dot_product)))
}

func (v *Vector) Normalize() {
	v.Scale(1 / v.Magnitude())
}

func Axpy(lhs *Vector, rhs *Vector, scalar float32) (Vector, error) {
	if !SameSize(lhs, rhs) {
		return Vector{}, errors.New("SizeMismatch: Vectors do not have the same size")
	}
	result := NewVector(make([]float32, lhs.Size))
	for i := 0; i < lhs.Size; i++ {
		result.Components[i] = lhs.Components[i]*scalar + rhs.Components[i]
	}
	return result, nil
}

func Dot(lhs *Vector, rhs *Vector) (float32, error) {
	if !SameSize(lhs, rhs) {
		return 0, errors.New("SizeMismatch: Vectors do not have the same size")
	}
	var result float32 = 0
	for i := 0; i < lhs.Size; i++ {
		result += lhs.Components[i] * rhs.Components[i]
	}
	return result, nil
}

func SameSize(lhs *Vector, rhs *Vector) bool {
	return lhs.Size == rhs.Size
}

func AddBorrow(lhs *Vector, rhs *Vector) (Vector, error) {
	if !SameSize(lhs, rhs) {
		return Vector{}, errors.New("SizeMismatch: Vectors do not have the same size")
	}
	result := NewVector(make([]float32, lhs.Size))
	for i := 0; i < lhs.Size; i++ {
		result.Components[i] = lhs.Components[i] + rhs.Components[i]
	}
	return result, nil
}

func SubBorrow(lhs *Vector, rhs *Vector) (Vector, error) {
	if !SameSize(lhs, rhs) {
		return Vector{}, errors.New("SizeMismatch: Vectors do not have the same size")
	}
	result := NewVector(make([]float32, lhs.Size))
	for i := 0; i < lhs.Size; i++ {
		result.Components[i] = lhs.Components[i] - rhs.Components[i]
	}
	return result, nil
}

func Add(lhs Vector, rhs Vector) (Vector, error) {
	return AddBorrow(&lhs, &rhs)
}

func Sub(lhs Vector, rhs Vector) (Vector, error) {
	return SubBorrow(&lhs, &rhs)
}

File: rust/src/vector.rs
use std::fmt;

// Define an error type
#[derive(Debug)]
pub enum MVectorError {
    SizeMismatch,
    OutOfBounds
}

impl fmt::Display for MVectorError {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        match self {
            MVectorError::SizeMismatch => write!(f, "SizeMismatch: Vectors do not have the same size"),
            MVectorError::OutOfBounds => write!(f, "OutOfBounds: index is out of bounds from Vector"),
        }
    }
}

pub mod mvec {
    use core::fmt;
    use vec_ops::dot;

    use super::MVectorError;

    #[derive(PartialEq)]
    pub struct Vector {
        components: Vec<f32>,
        size: usize
    }

    impl fmt::Display for Vector {
        fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
            write!(f, "{:?}", self.components)
        }
    }

    impl Vector {
        pub fn new(components: Vec<f32>) -> Self {
            Self { 
                size: components.len(),
                components, 
            }
        }

        pub fn scale(&mut self, scalar: f32){
            for component in self.components.iter_mut() {
                *component *= scalar;
            }
        }

        pub fn at(&self, index: usize) -> Result<f32, MVectorError> {
            if index >= self.size {
                return Err(MVectorError::OutOfBounds);
            }

            Ok(self.components[index])
        }

        pub fn at_ref(&self, index: usize) -> Result<&f32, MVectorError>{
            if index >= self.size {
                return Err(MVectorError::OutOfBounds);
            }

            Ok(&self.components[index])
        }

        pub fn magnitude(&self) -> f32 {
            f32::sqrt(dot(self, self).unwrap())
        }

        pub fn normalize(&mut self){
            self.scale(1.0 / self.magnitude());
        }

    }

    
    // ### Non standard operations on the `mvec::Vector` struct
    pub mod vec_ops {
        use std::ops;

        use crate::vector::MVectorError;
        use super::Vector;

        pub fn axpy(lhs: &Vector, rhs: &Vector, scalar: f32) -> Result<Vector, MVectorError>{
            if !same_size(lhs, rhs) {
                return Err(MVectorError::SizeMismatch);
            }

            let mut result: Vector = Vector::new(lhs.components.clone());
            result.scale(scalar);
            result.components.iter_mut()
                .zip(rhs.components.iter())
                .for_each(|(x, y)| *x += *y);

            Ok(result)
        }
        
        pub fn dot(lhs: &Vector, rhs: &Vector) -> Result<f32, MVectorError> {
            if !same_size(lhs, rhs) {
                return Err(MVectorError::SizeMismatch);
            }

            let mut result: f32 = 0.0;
            lhs.components.iter()
                .zip(rhs.components.iter())
                .for_each(|(x, y)| result += x * y);

            Ok(result)
        }

        pub fn same_size(lhs: &Vector, rhs: &Vector) -> bool {
            lhs.size == rhs.size
        }

        impl<'a, 'b> ops::Add<&'b Vector> for &'a Vector {
            type Output = Result<Vector, MVectorError>;

            fn add(self, rhs: &'b Vector) -> Self::Output {
                if !same_size(self, rhs) {
                    return Err(MVectorError::SizeMismatch);
                }
                let mut sum = Vector::new(self.components.clone());
                sum.components.iter_mut()
                    .zip(rhs.components.iter())
                    .for_each(|(x, y)| *x += *y);
                Ok(sum)
            }
        }

        impl<'a, 'b> ops::Sub<&'b Vector> for &'a Vector {
            type Output = Result<Vector, MVectorError>;
        
            fn sub(self, rhs: &'b Vector) -> Self::Output {
                if !same_size(self, rhs) {
                    return Err(MVectorError::SizeMismatch);
                }
                let mut difference = Vector::new(self.components.clone());
                difference.components.iter_mut()
                    .zip(rhs.components.iter())
                    .for_each(|(x, y)| *x -= *y);
                Ok(difference)
            }
        }

        impl ops::Add for Vector {
            type Output = Result<Vector, MVectorError>;
        
            fn add(self, rhs: Self) -> Self::Output { &self + &rhs }
        }

        impl ops::Sub for Vector {
            type Output = Result<Vector, MVectorError>;
            
            fn sub(self, rhs: Self) -> Self::Output { &self - &rhs }
        }
    }
}
